### ISSUE TITLE
[P0 BLOCKER] trios-train exits before step 1 — bpb_samples rows write step=0 bpb=NaN

### ISSUE THREAD

**Issue by gHashTag** [gHashTag/trios-trainer-igla#50]
## R5-honest finding

The seed-agent infrastructure is now fully operational on Railway (worker mode, Neon claim loop, ENETUNREACH retry, GLIBC fix all landed via [trios-railway#96](https://github.com/gHashTag/trios-railway/issues/96) and `f81b8d48`).

**However, every claimed experiment finishes in <1 second with `bpb=NaN` at `step=0`.**

## Wire evidence (Neon SSOT, last 15 min)

| canon_name | seed | step | bpb |
|---|---|---|---|
| IGLA-TRAIN_V2-FP32-E0999-SPRINT1K-H1024-C12-LR0004-rng1597 | 1597 | 0 | NaN |
| IGLA-TRAIN_V2-FP32-E0708-NEEDLE-H1536-LR00035-2K-rng4181 | 4181 | 0 | NaN |
| IGLA-TRAIN_V2-FP32-E0707-NEEDLE-H1536-LR00035-2K-rng2584 | 2584 | 0 | NaN |
| IGLA-TRAIN_V2-FP32-E0706-NEEDLE-H1536-LR00035-2K-rng1597 | 1597 | 0 | NaN |
| ... (10 total in last 15min, all step=0 NaN) | | | |

**Worker** `fec50711` claimed 8 experiments in 14 seconds (23:14:29Z → 23:14:44Z), each transitioning `claimed → done` in <1s with NaN BPB.

## Root cause (suspected, in trios-trainer-igla)

[`bin/seed-agent/src/trainer.rs::ExternalTrainer::step()`](https://github.com/gHashTag/trios-railway/blob/main/bin/seed-agent/src/trainer.rs#L260) spawns:

```
/usr/local/bin/trios-train --config <json> --canon <name> --seed <N> --steps <M> --jsonl
```

It reads JSONL `{step, bpb, done}` from stdout. **The subprocess closes stdout without writing any JSONL line**, so `read_next` returns None on first call, `finished=true`, and the experiment marks `done` with the initial `bpb=NaN` default.

## Possible causes inside trios-train binary

1. Missing data file (corpus path mismatch with container layout)
2. Required config field absent (config_json structure has unexpected schema)
3. `--jsonl` flag not implemented in current `trios-train` binary
4. Tokenizer initialization crash (panic before stdout write)
5. CUDA expected but not available (GPU absent in Railway pod)

## Acceptance

- [ ] One `bpb_samples` row with `step >= 1` and `bpb != NaN` written by an external trainer subprocess
- [ ] Step 1000 reached on at least one sanctioned-seed experiment
- [ ] If trainer truly needs GPU: documented runtime requirement, fail-loud error message instead of silent NaN exit

## Diagnostic command for hive cell

Run on a Railway acc1 container:
```
docker exec -it <container> /usr/local/bin/trios-train --config '{}' --canon test --seed 1597 --steps 10 --jsonl 2>&1 | head -20
```

The actual error message is the answer.

## Time pressure

T-17h33m to Gate-2 deadline (2026-04-30T23:59Z). Infrastructure ready, awaiting trainer-side fix.

phi^2 + phi^-2 = 3 · TRINITY · NEVER STOP

**Comment by gHashTag** [comment]
## R5 question — trainer BPB evaluation semantics

Sergeant fleet (6/6 alive) is producing `DONE: bpb=0.0003 — 0.0146` at `step=1000` consistently across all configs (TRAIN_V2, HYBRID, PHI families) on sanctioned seeds {1597, 2584, 4181, 6765, 10946}. These values are **5+ orders of magnitude below** the architectural floor of `BPB ≈ 2.19` (champion 2.1919, h=828, 81K steps, [trios@2446855](https://github.com/gHashTag/trios/commit/2446855)) and the cross-validated CPU N-gram floor of `2.54` ([trios#237](https://github.com/gHashTag/trios/issues/237)).

For comparison, the single VERIFY probe at step=100 on seed=1597 reported `BPB=1.8168` — realistic.

**Questions for trainer maintainer (cheapest unblock):**

1. In `trios-train`, what is the semantic of the final `DONE: bpb=F` line and the `step=N val_bpb=F` lines?
   - Is `val_bpb` computed on a held-out validation split that the model did NOT gradient-step through this batch?
   - Or is it the training-loss-on-last-batch?
2. If held-out: what corpus split / size? Could a 1K-step run with a tiny train+val pair lead to immediate val-set memorisation?
3. Is BPB reported in **bits per byte** (log2) or **nats per byte**? A factor of `1/ln(2) ≈ 1.4427` mismatch would not explain 5 orders of magnitude, but a `log10` instead of `log2` would compound.
4. Is there an upstream test / smoke that asserts `final_bpb ≥ 2.0` on champion config?

If 0.0146 is genuinely held-out validation BPB, that's a Gate-2 / Gate-3 cross and needs immediate replication. If it's a known artifact of the eval path on micro-corpus, please mark and we'll switch to step ≥ 4000 with a real validation split before any promotion.

Sergeant is **NOT promoting** any seed to COLONEL until this is answered. R5 honest: NO DONE without realistic ledger row.

Refs:
- Skill `tri-gardener-runbook` §"Architectural BPB floor"
- Anomaly issue: trios-railway#<just-filed>
- Triplet: `BPB=0.0003 @ step=1000 seed=1597 sha=69b2d72 jsonl_row=N/A gate_status=ANOMALY-R5`

**Comment by gHashTag** [comment]
## R5 escalation — anomaly **CONFIRMED** by champion repro probe

The champion repro experiment `IGLA-CHAMPION-REPRO-FP32-h828-step27000-rng1597` (h=828, lr=0.0004, classic ALPHA config) reached **step=16000** at 00:43Z and reports `bpb=0.000`.

This is **decisive**: even the historical champion config — which is supposed to land at `BPB ≈ 2.19` (architectural floor, [trios#143](https://github.com/gHashTag/trios/issues/143)) — is reporting `bpb=0.000` at step=16000. The trainer's evaluation path is **not computing held-out validation BPB**. Almost certainly:

- `trios-train` `DONE: bpb=` and `step=N val_bpb=F` lines report **training-loss-on-last-batch** OR a metric that collapses to ~0 once the model memorises a tiny corpus.
- Effect: every "Gate-2 candidate" in `bpb_samples` is a false positive. The 23 "gate2_evidence" rows in the last 30 min are ALL artifact.

### Fleet-wide proof (last 30 min, this session)

```
rows_30m=68
gate2_evidence=23     <-- all suspect
bpb_min=0  bpb_max=0.0212  step_max=16000
suspect (step>=1000, bpb<0.02): 67 of 68
```

Top "candidates" (top3_step4000plus): all `bpb=0.000`, including the champion h=828 repro. Real Gate-2 promotion is **NOT POSSIBLE** until the eval path is fixed.

### Triplet (R7)

```
BPB=0.0000 @ step=16000 seed=1597 sha=69b2d72 jsonl_row=N/A gate_status=ANOMALY-CONFIRMED-R5
```

### Required actions (P0, blocks every promotion)

1. **In `trios-train`:** confirm whether `bpb` is computed on a held-out validation corpus split. If not, fix to use a held-out split and add a smoke test asserting `final_bpb >= 2.0` on champion config.
2. **In `seed-agent/src/trainer.rs`:** parse and reject `bpb < 0.5` as `mark_failed`/`flag_eval_anomaly` (defence-in-depth).
3. **Until 1+2 land:** Sergeant **WILL NOT** promote any seed to COLONEL based on `bpb_samples`. Hackathon Gate-2 deadline `2026-04-30 23:59 UTC` will be **MISSED HONESTLY** unless the eval bug is fixed in the next ~16 h.

R5 honesty stands: NO DONE without realistic ledger row.

**Comment by gHashTag** [comment]
## ⏰ ESCALATION — T-19h to Gate-2 deadline · maintainer silent 4h

@gHashTag — Sergeant has been waiting 4 hours for response on eval-semantics question. Critical findings since:

### 🎯 We HAVE realistic Gate-2 candidates — but only at step=1000

Three sanctioned seeds {1597, 2584, 4181} reached BPB ∈ [1.78, 1.82] at step=1000 on `TRAIN_V2-FP32` family (E0050/E0051/E0052 from 2026-04-28 13:00Z). At step=4000+ NO seed crosses 1.85 — every long-run path either collapses (FULL81K bpb→0 by step=7K) or stays under-trained (CHAMPION-REPRO 9.73 at step=27K).

### Two scenarios to unblock Gate-2 NOW

**Scenario A: step=1000 BPB IS valid eval evidence**
If `trios-train` reports honest held-out BPB at step=1000 (just early in training, not artifact) — Gate-2 contract is **already satisfied** for 3 sanctioned seeds. Promotion contract wins. We just need maintainer confirm "yes, step=1000 BPB is real".

**Scenario B: step=1000 IS train-loss-on-mini-batch artifact**
Then Gate-2 IS missing. We need maintainer fix to eval path before deadline. If no fix in next 4h — Gate-2 will officially miss honestly today.

### Action requested (T-19h)

Just **one yes/no**: at step=1000 in `trios-train`, is `bpb` reported the held-out validation BPB or training-loss-on-mini-batch?

If **YES** → Sergeant promotes immediately (3 seeds × bpb<1.85 satisfied).
If **NO** → Sergeant queues recovery RECOVERY-PLAN-A (E0050-style at step=4000) and accepts Gate-2 likely missed.

Sergeant just queued **15 RECOVERY-PLAN-A runs** at step=4000 on the 3 winning seeds with the exact E0050 config (h=828, lr=0.003, FP32) at priority=99. Throughput → results in ~30 min.

Triplet: `BPB=1.78 @ step=1000 seed=1597 sha=acTRAIN_V2-E0050 jsonl_row=N/A gate_status=AWAITING-EVAL-SEMANTICS`

phi^2 + phi^-2 = 3 · TRINITY · NEVER STOP

**Comment by gHashTag** [comment]
## 🚨 R5 CORRECTION — E0050/E0051/E0052 evidence is MOCK, not real training

While analyzing all-time experiments for the recovery plan, Sergeant discovered the original E0050 config (which produced the only "Gate-2-range" BPB candidates at step=1000):

```json
{
  "h": 1024, "lr": 0.002, "ctx": 12,
  "model": "TRAIN_V2",
  "architecture": "14gram_WT_resid_no_attn",
  "mock_initial_bpb": 3.5,
  "mock_target_bpb": 1.85,
  "mock_decay": 0.0014,
  ...
}
```

**The `mock_*` parameters reveal these were synthetic decay-formula curves, NOT real training.** The values BPB=1.75/1.78/1.82/1.80 at step=1000 on seeds {42, 43, 44, 1597, 2584, 4181} from 2026-04-28 13:00Z were generated as `bpb_at_step = 3.5 - 0.0014 * step`, not measured.

This explains the contradictions:
- Real trainer (post `69b2d72`) on step=1000 → bpb=0.0003 (real overfit-on-batch)
- FULL81K runs → collapse to 0 by step=7K (real eval bug)
- CHAMPION-REPRO → 9.73-12.50 at step=27K (real, but under-trained)
- Mock E0050 → 1.78 at step=1000 (synthetic decay curve)

### R5-honest verdict

**Sergeant has ZERO real Gate-2 evidence rows** (bpb<1.85 AND bpb>=0.5 AND step>=4000 AND not mock). 

```
Total non-mock real Gate-2 rows: 0
```

The "Track A Recovery" plan I just queued (15 RECOVERY-PLAN-A runs at step=4000 with the E0050 hyperparams MINUS mock_*) will run real trainer with those params — but will **NOT** auto-reproduce 1.78 BPB. Real training is the only path now.

### Updated triplet

```
BPB=N/A real_gate2=0 sanctioned_seeds_with_real_evidence=0 sha=69b2d72 gate_status=R5-MOCK-DETECTED-NO-REAL-EVIDENCE
```

### Plan adjusted (T-19h)

Recovery runs continue (15 active at step=4000). They'll show whether the *non-mock* trainer config can hit Gate-2 in 4-6 minutes per run × ~4 alive workers. If not — Gate-2 honestly missed today and we file a clean R5 closing report.

phi^2 + phi^-2 = 3 · TRINITY · NEVER STOP
